serialize.go

```diff
@@ -0,0 +1,153 @@
+//
+// If copyToSqlite is true, the data will be copied. This should be set to true
+// if this will be used with SQLITE_DESERIALIZE_FREEONCLOSE.
+func NewSerialized(schema string, data []byte, copyToSqlite bool) *Serialized {
+	s := &Serialized{
+		schema: schema,
+		data:   data,
+	}
+	if copyToSqlite {
+		sqliteData := (*[1 << 28]uint8)(unsafe.Pointer(C.sqlite3_malloc(C.int(len(data)))))[:len(data):len(data)]
+		copy(sqliteData, data)
+		s.data = sqliteData
+		s.shouldFreeData = true
+		s.sqliteOwnsData = true
+		runtime.SetFinalizer(s, func(s *Serialized) { s.free() })
+	}
+	return s
+}
+
+// Schema returns the schema for this serialized DB.
+func (s *Serialized) Schema() string {
+	if s.schema == "" {
+		return "main"
+	}
+	return s.schema
+}
+
+// Bytes returns the serialized bytes. Do not mutate this value. This is only
+// valid for the life of its receiver and should be copied for any other
+// longer-term use.
+func (s *Serialized) Bytes() []byte { return s.data }
+
+func (s *Serialized) free() {
+	if len(s.data) > 0 && s.shouldFreeData {
+		s.shouldFreeData = false
+		s.sqliteOwnsData = false
+		C.sqlite3_free(unsafe.Pointer(&s.data[0]))
+	}
+	s.data = nil
+}
+
+// SerializeFlags are flags used for Serialize.
+type SerializeFlags int
+
+const (
+	SQLITE_SERIALIZE_NOCOPY SerializeFlags = C.SQLITE_SERIALIZE_NOCOPY
+)
+
+// Serialize serializes the given schema. Returns nil on error. If
+// SQLITE_SERIALIZE_NOCOPY flag is set, the data may only be valid as long as
+// the database.
+//
+// https://www.sqlite.org/c3ref/serialize.html
+func (conn *Conn) Serialize(schema string, flags ...SerializeFlags) *Serialized {
+	var serializeFlags SerializeFlags
+	for _, f := range flags {
+		serializeFlags |= f
+	}
+
+	cschema := cmain
+	if schema != "" && schema != "main" {
+		cschema = C.CString(schema)
+		defer C.free(unsafe.Pointer(cschema))
+	}
+
+	var csize C.sqlite3_int64
+	res := C.sqlite3_serialize(conn.conn, cschema, &csize, C.uint(serializeFlags))
+	if res == nil {
+		return nil
+	}
+
+	s := &Serialized{
+		schema:         schema,
+		data:           (*[1 << 28]uint8)(unsafe.Pointer(res))[:csize:csize],
+		sqliteOwnsData: true,
+	}
+	// Free the memory only if they didn't specify nocopy
+	if serializeFlags&SQLITE_SERIALIZE_NOCOPY == 0 {
+		s.shouldFreeData = true
+		runtime.SetFinalizer(s, func(s *Serialized) { s.free() })
+	}
+	return s
+}
+
+// DeserializeFlags are flags used for Deserialize.
+type DeserializeFlags int
+
+const (
+	SQLITE_DESERIALIZE_FREEONCLOSE DeserializeFlags = C.SQLITE_DESERIALIZE_FREEONCLOSE
+	SQLITE_DESERIALIZE_RESIZEABLE  DeserializeFlags = C.SQLITE_DESERIALIZE_RESIZEABLE
+	SQLITE_DESERIALIZE_READONLY    DeserializeFlags = C.SQLITE_DESERIALIZE_READONLY
+)
+
+// Reopens the database as in-memory representation of given serialized bytes.
+// The given *Serialized instance should remain referenced (i.e. not GC'd) for
+// the life of the DB since the bytes within are referenced directly.
```

Comment:
This is not something the caller can guarantee, short of importing [go4.org/unsafe/assume-no-moving-gc](https://go4.org/unsafe/assume-no-moving-gc).

Generally, it's probably not a good idea to let C and Go memory cross the boundaries like this. It adds complexity including the need for the finalizers, and the invalid state of Serialized values after being used.

In Serialize, I'd always make a copy into Go memory unless SQLITE_SERIALIZE_NOCOPY is set. That way there is never a need for freeing the memory. If an application can't stand copies for performance reasons, they can use SQLITE_SERIALIZE_NOCOPY. Otherwise, they can probably tolerate two as well as one.

In Deserialize, I would always copy the input into sqlite3_malloc64'd memory, and force SQLITE_DESERIALIZE_FREEONCLOSE on. This is unfortunate in terms of memory usage, especially when using //go:embed which is my use case, and basically precludes the use of the szBuf > szDb feature, but it is the only safe API because we can't rely on Go memory staying still.

The good news is that this removes any need for the Serialized type.

Reply:
Hrmm, yeah all my machinations here were to avoid copies, but it sounds like I can't have a stable pointer. So I guess we have to have copies.

I did this for a side project (https://github.com/cretz/temporal-sdk-go-advanced/tree/main/temporalsqlite for SQLite on https://temporal.io/) so it may be a few days until I can get to simplifying. If needing sooner or you have the bandwidth, feel free to steal the impl here and simplify w/ always-copy.

Reply:
Oh, actually it might be worth waiting for https://go.dev/issue/46787. I wonder if we can provide the same API I describe above, and implement it with copies in Go 1.18, and with Pinner in Go 1.19.

Reply:
https://github.com/golang/go/issues/46787 has landed.